File: data/src/main/java/org/cryptomator/data/cloud/crypto/CryptoImplVaultFormatPre7.java
package org.cryptomator.data.cloud.crypto;

import android.content.Context;

import org.apache.commons.codec.binary.Base32;
import org.apache.commons.codec.binary.BaseNCodec;
import org.cryptomator.cryptolib.Cryptors;
import org.cryptomator.cryptolib.api.AuthenticationFailedException;
import org.cryptomator.cryptolib.api.Cryptor;
import org.cryptomator.cryptolib.common.MessageDigestSupplier;
import org.cryptomator.domain.CloudFile;
import org.cryptomator.domain.CloudFolder;
import org.cryptomator.domain.CloudNode;
import org.cryptomator.domain.exception.AlreadyExistException;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.EmptyDirFileException;
import org.cryptomator.domain.exception.NoSuchCloudFileException;
import org.cryptomator.domain.repository.CloudContentRepository;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.domain.usecases.cloud.ByteArrayDataSource;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.cryptomator.domain.usecases.cloud.UploadState;
import org.cryptomator.util.Optional;
import org.cryptomator.util.Supplier;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import timber.log.Timber;

import static org.cryptomator.domain.usecases.ProgressAware.NO_OP_PROGRESS_AWARE;
import static org.cryptomator.util.Encodings.UTF_8;

final class CryptoImplVaultFormatPre7 extends CryptoImplDecorator {

	static final int SHORTENING_THRESHOLD = 129;
	private static final String DIR_PREFIX = "0";
	private static final String SYMLINK_PREFIX = "1S";
	private static final String LONG_NAME_FILE_EXT = ".lng";
	private static final String METADATA_DIR_NAME = "m";
	private static final BaseNCodec BASE32 = new Base32();
	private static final Pattern BASE32_ENCRYPTED_NAME_PATTERN = Pattern.compile("^(0|1S)?(([A-Z2-7]{8})*[A-Z2-7=]{8})$");

	CryptoImplVaultFormatPre7(Context context, Supplier<Cryptor> cryptor, CloudContentRepository cloudContentRepository, CloudFolder storageLocation, DirIdCache dirIdCache) {
		super(context, cryptor, cloudContentRepository, storageLocation, dirIdCache, SHORTENING_THRESHOLD);
	}

	@Override
	CryptoFolder folder(CryptoFolder cryptoParent, String cleartextName) throws BackendException {
		String dirFileName = encryptFolderName(cryptoParent, cleartextName);
		CloudFile dirFile = cloudContentRepository.file(dirIdInfo(cryptoParent).getCloudFolder(), dirFileName);
		return folder(cryptoParent, cleartextName, dirFile);
	}

	@Override
	CryptoFolder create(CryptoFolder folder) throws BackendException {
		assertCryptoFolderAlreadyExists(folder);
		DirIdCache.DirIdInfo dirIdInfo = dirIdInfo(folder);
		CloudFolder createdCloudFolder = cloudContentRepository.create(dirIdInfo.getCloudFolder());
		byte[] dirId = dirIdInfo.getId().getBytes(UTF_8);
		CloudFile createdDirFile = cloudContentRepository.write(folder.getDirFile(), ByteArrayDataSource.from(dirId), NO_OP_PROGRESS_AWARE, false, dirId.length);
		CryptoFolder result = folder(folder, createdDirFile);
		addFolderToCache(result, dirIdInfo.withCloudFolder(createdCloudFolder));
		return result;
	}

	@Override
	String encryptName(CryptoFolder cryptoParent, String name) throws BackendException {
		return encryptName(cryptoParent, name, "");
	}

	private String encryptName(CryptoFolder cryptoParent, String name, String prefix) throws BackendException {
		String ciphertextName = prefix + cryptor().fileNameCryptor().encryptFilename(name, dirIdInfo(cryptoParent).getId().getBytes(UTF_8));
		if (ciphertextName.length() > shorteningThreshold) {
			ciphertextName = deflate(ciphertextName);
		}
		return ciphertextName;
	}

	private String deflate(String longFileName) throws BackendException {
		byte[] longFilenameBytes = longFileName.getBytes(UTF_8);
		byte[] hash = MessageDigestSupplier.SHA1.get().digest(longFilenameBytes);
		String shortFileName = BASE32.encodeAsString(hash) + LONG_NAME_FILE_EXT;
		CloudFile metadataFile = metadataFile(shortFileName);
		byte[] data = longFileName.getBytes(UTF_8);
		try {
			cloudContentRepository.create(metadataFile.getParent());
		} catch (AlreadyExistException e) {
		}
		cloudContentRepository.write(metadataFile, ByteArrayDataSource.from(data), NO_OP_PROGRESS_AWARE, true, data.length);
		return shortFileName;
	}

	private String inflate(String shortFileName) throws BackendException {
		CloudFile metadataFile = metadataFile(shortFileName);
		ByteArrayOutputStream out = new ByteArrayOutputStream();
		cloudContentRepository.read(metadataFile, Optional.empty(), out, NO_OP_PROGRESS_AWARE);
		return new String(out.toByteArray(), UTF_8);
	}

	private CloudFile inflatePermanently(CloudFile cloudFile, String longFileName) throws BackendException {
		Timber.tag("CryptoFs").i("inflatePermanently: %s -> %s", cloudFile.getName(), longFileName);
		CloudFile newCiphertextFile = cloudContentRepository.file(cloudFile.getParent(), longFileName);
		cloudContentRepository.move(cloudFile, newCiphertextFile);
		return newCiphertextFile;
	}

	private CloudFile metadataFile(String shortFilename) throws BackendException {
		CloudFolder firstLevelFolder = cloudContentRepository.folder(metadataFolder(), shortFilename.substring(0, 2));
		CloudFolder secondLevelFolder = cloudContentRepository.folder(firstLevelFolder, shortFilename.substring(2, 4));
		return cloudContentRepository.file(secondLevelFolder, shortFilename);
	}

	private CloudFolder metadataFolder() throws BackendException {
		return cloudContentRepository.folder(storageLocation(), METADATA_DIR_NAME);
	}

	@Override
	List<CryptoNode> list(CryptoFolder cryptoFolder) throws BackendException {
		DirIdCache.DirIdInfo dirIdInfo = dirIdInfo(cryptoFolder);
		String dirId = dirIdInfo(cryptoFolder).getId();
		CloudFolder lvl2Dir = dirIdInfo.getCloudFolder();
		List<CloudNode> ciphertextNodes = cloudContentRepository.list(lvl2Dir);
		List<CryptoNode> result = new ArrayList<>();
		for (CloudNode node : ciphertextNodes) {
			if (node instanceof CloudFile) {
				ciphertextToCleartextNode(cryptoFolder, dirId, node).ifPresent(result::add);
			}
		}
		return result;
	}

	private Optional<CryptoNode> ciphertextToCleartextNode(CryptoFolder cryptoFolder, String dirId, CloudNode cloudNode) throws BackendException {
		CloudFile cloudFile = (CloudFile) cloudNode;
		String ciphertextName = cloudFile.getName();
		if (ciphertextName.endsWith(LONG_NAME_FILE_EXT)) {
			try {
				ciphertextName = inflate(ciphertextName);
				if (ciphertextName.length() <= shorteningThreshold) {
					cloudFile = inflatePermanently(cloudFile, ciphertextName);
				}
			} catch (NoSuchCloudFileException e) {
				Timber.tag("CryptoFs").e("Missing mFile: %s", ciphertextName);
				return Optional.empty();
			} catch (BackendException e) {
				Timber.tag("CryptoFs").e(e, "Failed to read mFile: %s", ciphertextName);
				return Optional.empty();
			}
		}
		String cleartextName;
		try {
			cleartextName = decryptName(dirId, ciphertextName.toUpperCase());
		} catch (AuthenticationFailedException e) {
			Timber.tag("CryptoFs").w("File name authentication failed: %s", cloudFile.getPath());
			return Optional.empty();
		} catch (IllegalArgumentException e) {
			Timber.tag("CryptoFs").d("Illegal ciphertext filename: %s", cloudFile.getPath());
			return Optional.empty();
		}
		if (cleartextName == null || ciphertextName.startsWith(SYMLINK_PREFIX)) {
			return Optional.empty();
		} else if (ciphertextName.startsWith(DIR_PREFIX)) {
			return Optional.of(folder(cryptoFolder, cleartextName, cloudFile));
		} else {
			Optional<Long> cleartextSize = Optional.empty();
			if (cloudFile.getSize().isPresent()) {
				long ciphertextSizeWithoutHeader = cloudFile.getSize().get() - cryptor().fileHeaderCryptor().headerSize();
				if (ciphertextSizeWithoutHeader >= 0) {
					cleartextSize = Optional.of(Cryptors.cleartextSize(ciphertextSizeWithoutHeader, cryptor()));
				}
			}
			return Optional.of(file(cryptoFolder, cleartextName, cloudFile, cleartextSize));
		}
	}

	@Override
	String decryptName(String dirId, String encryptedName) {
		Optional<String> ciphertextName = extractEncryptedName(encryptedName);
		if (ciphertextName.isPresent()) {
			return cryptor().fileNameCryptor().decryptFilename(ciphertextName.get(), dirId.getBytes(UTF_8));
		} else {
			return null;
		}
	}

	@Override
	Optional<String> extractEncryptedName(String ciphertextName) {
		Matcher matcher = BASE32_ENCRYPTED_NAME_PATTERN.matcher(ciphertextName);
		if (matcher.find(0)) {
			return Optional.of(matcher.group(2));
		} else {
			return Optional.empty();
		}
	}

	@Override
	CryptoSymlink symlink(CryptoFolder cryptoParent, String cleartextName, String target) throws BackendException {
		String ciphertextName = encryptSymlinkName(cryptoParent, cleartextName);
		CloudFile cloudFile = cloudContentRepository.file(dirIdInfo(cryptoParent).getCloudFolder(), ciphertextName);
		return new CryptoSymlink(cryptoParent, cleartextName, path(cryptoParent, cleartextName), target, cloudFile);
	}

	private String encryptSymlinkName(CryptoFolder cryptoFolder, String name) throws BackendException {
		return encryptName(cryptoFolder, name, SYMLINK_PREFIX);
	}

	@Override
	String encryptFolderName(CryptoFolder cryptoFolder, String name) throws BackendException {
		return encryptName(cryptoFolder, name, DIR_PREFIX);
	}

	@Override
	CryptoFolder move(CryptoFolder source, CryptoFolder target) throws BackendException {
		assertCryptoFolderAlreadyExists(target);
		CryptoFolder result = folder(target.getParent(), target.getName(), cloudContentRepository.move(source.getDirFile(), target.getDirFile()));

		evictFromCache(source);
		evictFromCache(target);
		return result;
	}

	@Override
	CryptoFile move(CryptoFile source, CryptoFile target) throws BackendException {
		assertCryptoFileAlreadyExists(target);
		return file(target, cloudContentRepository.move(source.getCloudFile(), target.getCloudFile()), source.getSize());
	}

	@Override
	void delete(CloudNode node) throws BackendException {
		if (node instanceof CryptoFolder) {
			CryptoFolder cryptoFolder = (CryptoFolder) node;
			List<CryptoFolder> cryptoSubfolders = deepCollectSubfolders(cryptoFolder);
			for (CryptoFolder cryptoSubfolder : cryptoSubfolders) {
				cloudContentRepository.delete(dirIdInfo(cryptoSubfolder).getCloudFolder());
			}
			cloudContentRepository.delete(dirIdInfo(cryptoFolder).getCloudFolder());
			cloudContentRepository.delete(cryptoFolder.getDirFile());
			evictFromCache(cryptoFolder);
		} else if (node instanceof CryptoFile) {
			CryptoFile cryptoFile = (CryptoFile) node;
			cloudContentRepository.delete(cryptoFile.getCloudFile());
		}
	}

	@Override
	String loadDirId(CryptoFolder folder) throws BackendException, EmptyDirFileException {
		if (RootCryptoFolder.isRoot(folder)) {
			return CryptoConstants.ROOT_DIR_ID;
		} else if (cloudContentRepository.exists(folder.getDirFile())) {
			return new String(loadContentsOfDirFile(folder), UTF_8);
		} else {
			return newDirId();
		}
	}

	@Override
	DirIdCache.DirIdInfo createDirIdInfo(CryptoFolder folder) throws BackendException {
		String dirId = loadDirId(folder);
		return dirIdCache.put(folder, createDirIdInfoFor(dirId));
	}

	@Override
	public CryptoFile write(CryptoFile cryptoFile, DataSource data, ProgressAware<UploadState> progressAware, boolean replace, long length) throws BackendException {
		return writeShortNameFile(cryptoFile, data, progressAware, replace, length);
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/crypto/CryptoImplVaultFormatPre7.kt
package org.cryptomator.data.cloud.crypto

import android.content.Context
import com.google.common.io.BaseEncoding
import org.apache.commons.codec.binary.Base32
import org.apache.commons.codec.binary.BaseNCodec
import org.cryptomator.cryptolib.api.AuthenticationFailedException
import org.cryptomator.cryptolib.api.Cryptor
import org.cryptomator.cryptolib.common.MessageDigestSupplier
import org.cryptomator.data.cloud.crypto.DirIdCache.DirIdInfo
import org.cryptomator.data.cloud.crypto.RootCryptoFolder.Companion.isRoot
import org.cryptomator.domain.Cloud
import org.cryptomator.domain.CloudFile
import org.cryptomator.domain.CloudFolder
import org.cryptomator.domain.CloudNode
import org.cryptomator.domain.exception.AlreadyExistException
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.EmptyDirFileException
import org.cryptomator.domain.exception.NoSuchCloudFileException
import org.cryptomator.domain.exception.ParentFolderIsNullException
import org.cryptomator.domain.repository.CloudContentRepository
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.domain.usecases.cloud.ByteArrayDataSource.Companion.from
import org.cryptomator.domain.usecases.cloud.DataSource
import org.cryptomator.domain.usecases.cloud.UploadState
import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets
import java.util.function.Supplier
import java.util.regex.Pattern
import kotlin.streams.toList
import timber.log.Timber

internal class CryptoImplVaultFormatPre7(
	context: Context,
	cryptor: Supplier<Cryptor>,
	cloudContentRepository: CloudContentRepository<Cloud, CloudNode, CloudFolder, CloudFile>,
	storageLocation: CloudFolder,
	dirIdCache: DirIdCache
) :
	CryptoImplDecorator(
		context, cryptor, cloudContentRepository, storageLocation, dirIdCache, SHORTENING_THRESHOLD
	) {

	@Throws(BackendException::class)
	override fun folder(cryptoParent: CryptoFolder, cleartextName: String): CryptoFolder {
		val dirFileName = encryptFolderName(cryptoParent, cleartextName)
		val dirFile = cloudContentRepository.file(dirIdInfo(cryptoParent).cloudFolder, dirFileName)
		return folder(cryptoParent, cleartextName, dirFile)
	}

	@Throws(BackendException::class)
	override fun create(folder: CryptoFolder): CryptoFolder {
		requireNotNull(folder.dirFile)
		assertCryptoFolderAlreadyExists(folder)
		val dirIdInfo = dirIdInfo(folder)
		val createdCloudFolder = cloudContentRepository.create(dirIdInfo.cloudFolder)
		val dirId = dirIdInfo.id.toByteArray(StandardCharsets.UTF_8)
		val createdDirFile = cloudContentRepository.write(folder.dirFile, from(dirId), ProgressAware.NO_OP_PROGRESS_AWARE_UPLOAD, false, dirId.size.toLong())
		return folder(folder, createdDirFile).also {
			addFolderToCache(it, dirIdInfo.withCloudFolder(createdCloudFolder))
		}
	}

	@Throws(BackendException::class)
	override fun encryptName(cryptoParent: CryptoFolder, name: String): String {
		return encryptName(cryptoParent, name, "")
	}

	@Throws(BackendException::class)
	private fun encryptName(cryptoParent: CryptoFolder, name: String, prefix: String): String {
		var ciphertextName = prefix + cryptor().fileNameCryptor().encryptFilename(BaseEncoding.base32(), name, dirIdInfo(cryptoParent).id.toByteArray(StandardCharsets.UTF_8))
		if (ciphertextName.length > shorteningThreshold) {
			ciphertextName = deflate(ciphertextName)
		}
		return ciphertextName
	}

	@Throws(BackendException::class)
	private fun deflate(longFileName: String): String {
		val longFilenameBytes = longFileName.toByteArray(StandardCharsets.UTF_8)
		val hash = MessageDigestSupplier.SHA1.get().digest(longFilenameBytes)
		val shortFileName = BASE32.encodeAsString(hash) + LONG_NAME_FILE_EXT
		val metadataFile = metadataFile(shortFileName)
		val data = longFileName.toByteArray(StandardCharsets.UTF_8)
		try {
			cloudContentRepository.create(metadataFile.parent)
		} catch (e: AlreadyExistException) {
		}
		cloudContentRepository.write(metadataFile, from(data), ProgressAware.NO_OP_PROGRESS_AWARE_UPLOAD, true, data.size.toLong())
		return shortFileName
	}

	@Throws(BackendException::class)
	private fun inflate(shortFileName: String): String {
		val metadataFile = metadataFile(shortFileName)
		val out = ByteArrayOutputStream()
		cloudContentRepository.read(metadataFile, null, out, ProgressAware.NO_OP_PROGRESS_AWARE_DOWNLOAD)
		return String(out.toByteArray(), StandardCharsets.UTF_8)
	}

	@Throws(BackendException::class)
	private fun inflatePermanently(cloudFile: CloudFile, longFileName: String): CloudFile {
		Timber.tag("CryptoFs").i("inflatePermanently: %s -> %s", cloudFile.name, longFileName)
		val newCiphertextFile = cloudContentRepository.file(cloudFile.parent, longFileName)
		cloudContentRepository.move(cloudFile, newCiphertextFile)
		return newCiphertextFile
	}

	@Throws(BackendException::class)
	private fun metadataFile(shortFilename: String): CloudFile {
		val firstLevelFolder = cloudContentRepository.folder(metadataFolder(), shortFilename.substring(0, 2))
		val secondLevelFolder = cloudContentRepository.folder(firstLevelFolder, shortFilename.substring(2, 4))
		return cloudContentRepository.file(secondLevelFolder, shortFilename)
	}

	@Throws(BackendException::class)
	private fun metadataFolder(): CloudFolder {
		return cloudContentRepository.folder(storageLocation(), METADATA_DIR_NAME)
	}

	@Throws(BackendException::class)
	override fun list(cryptoFolder: CryptoFolder): List<CryptoNode> {
		val dirIdInfo = dirIdInfo(cryptoFolder)
		val dirId = dirIdInfo(cryptoFolder).id
		val lvl2Dir = dirIdInfo.cloudFolder
		return cloudContentRepository
			.list(lvl2Dir)
			.filterIsInstance<CloudFile>()
			.parallelStream()
			.map { node ->
				ciphertextToCleartextNode(cryptoFolder, dirId, node)
			}
			.toList()
			.filterNotNull()
	}

	@Throws(BackendException::class)
	private fun ciphertextToCleartextNode(cryptoFolder: CryptoFolder, dirId: String, cloudNode: CloudFile): CryptoNode? {
		var cloudFile = cloudNode
		var ciphertextName = cloudFile.name
		if (ciphertextName.endsWith(LONG_NAME_FILE_EXT)) {
			try {
				ciphertextName = inflate(ciphertextName)
				if (ciphertextName.length <= shorteningThreshold) {
					cloudFile = inflatePermanently(cloudFile, ciphertextName)
				}
			} catch (e: NoSuchCloudFileException) {
				Timber.tag("CryptoFs").e("Missing mFile: %s", ciphertextName)
				return null
			} catch (e: BackendException) {
				Timber.tag("CryptoFs").e(e, "Failed to read mFile: %s", ciphertextName)
				return null
			}
		}
		val cleartextName: String? = try {
			decryptName(dirId, ciphertextName.uppercase())
		} catch (e: AuthenticationFailedException) {
			Timber.tag("CryptoFs").w("File name authentication failed: %s", cloudFile.path)
			return null
		} catch (e: IllegalArgumentException) {
			Timber.tag("CryptoFs").d("Illegal ciphertext filename: %s", cloudFile.path)
			return null
		}
		return if (cleartextName == null || ciphertextName.startsWith(SYMLINK_PREFIX)) {
			null
		} else if (ciphertextName.startsWith(DIR_PREFIX)) {
			folder(cryptoFolder, cleartextName, cloudFile)
		} else {
			val cleartextSize = cloudFile.size?.let {
				val ciphertextSizeWithoutHeader: Long = it - cryptor().fileHeaderCryptor().headerSize()
				if (ciphertextSizeWithoutHeader >= 0) {
					cryptor().fileContentCryptor().cleartextSize(ciphertextSizeWithoutHeader)
				} else {
					null
				}
			}
			file(cryptoFolder, cleartextName, cloudFile, cleartextSize)
		}
	}

	override fun decryptName(dirId: String, encryptedName: String): String? {
		val ciphertextName = extractEncryptedName(encryptedName)
		return if (ciphertextName != null) {
			cryptor().fileNameCryptor().decryptFilename(BaseEncoding.base32(), ciphertextName, dirId.toByteArray(StandardCharsets.UTF_8))
		} else {
			null
		}
	}

	override fun extractEncryptedName(ciphertextName: String): String? {
		val matcher = BASE32_ENCRYPTED_NAME_PATTERN.matcher(ciphertextName)
		return if (matcher.find(0)) {
			matcher.group(2)
		} else {
			null
		}
	}

	@Throws(BackendException::class)
	override fun symlink(cryptoParent: CryptoFolder, cleartextName: String, target: String): CryptoSymlink {
		val ciphertextName = encryptSymlinkName(cryptoParent, cleartextName)
		val cloudFile = cloudContentRepository.file(dirIdInfo(cryptoParent).cloudFolder, ciphertextName)
		return CryptoSymlink(cryptoParent, cleartextName, path(cryptoParent, cleartextName), target, cloudFile)
	}

	@Throws(BackendException::class)
	private fun encryptSymlinkName(cryptoFolder: CryptoFolder, name: String): String {
		return encryptName(cryptoFolder, name, SYMLINK_PREFIX)
	}

	@Throws(BackendException::class)
	override fun encryptFolderName(cryptoFolder: CryptoFolder, name: String): String {
		return encryptName(cryptoFolder, name, DIR_PREFIX)
	}

	@Throws(BackendException::class)
	override fun move(source: CryptoFolder, target: CryptoFolder): CryptoFolder {
		requireNotNull(source.dirFile)
		requireNotNull(target.dirFile)
		target.parent?.let {
			assertCryptoFolderAlreadyExists(target)
			return folder(it, target.name, cloudContentRepository.move(source.dirFile, target.dirFile)).also {
				evictFromCache(source)
				evictFromCache(target)
			}
		} ?: throw ParentFolderIsNullException(target.name)
	}

	@Throws(BackendException::class)
	override fun move(source: CryptoFile, target: CryptoFile): CryptoFile {
		assertCryptoFileAlreadyExists(target)
		return file(target, cloudContentRepository.move(source.cloudFile, target.cloudFile), source.size)
	}

	@Throws(BackendException::class)
	override fun delete(node: CloudNode) {
		if (node is CryptoFolder) {
			requireNotNull(node.dirFile)
			val cryptoSubfolders = deepCollectSubfolders(node)
			for (cryptoSubfolder in cryptoSubfolders) {
				cloudContentRepository.delete(dirIdInfo(cryptoSubfolder).cloudFolder)
			}
			cloudContentRepository.delete(dirIdInfo(node).cloudFolder)
			cloudContentRepository.delete(node.dirFile)
			evictFromCache(node)
		} else if (node is CryptoFile) {
			cloudContentRepository.delete(node.cloudFile)
		}
	}

	@Throws(BackendException::class, EmptyDirFileException::class)
	override fun loadDirId(folder: CryptoFolder): String {
		return if (isRoot(folder)) {
			CryptoConstants.ROOT_DIR_ID
		} else if (folder.dirFile != null && cloudContentRepository.exists(folder.dirFile)) {
			String(loadContentsOfDirFile(folder), StandardCharsets.UTF_8)
		} else {
			newDirId()
		}
	}

	@Throws(BackendException::class)
	override fun createDirIdInfo(folder: CryptoFolder): DirIdInfo {
		val dirId = loadDirId(folder)
		return dirIdCache.put(folder, createDirIdInfoFor(dirId))
	}

	@Throws(BackendException::class)
	override fun write(cryptoFile: CryptoFile, data: DataSource, progressAware: ProgressAware<UploadState>, replace: Boolean, length: Long): CryptoFile {
		return writeShortNameFile(cryptoFile, data, progressAware, replace, length)
	}

	companion object {

		const val SHORTENING_THRESHOLD = 129
		private const val DIR_PREFIX = "0"
		private const val SYMLINK_PREFIX = "1S"
		private const val LONG_NAME_FILE_EXT = ".lng"
		private const val METADATA_DIR_NAME = "m"
		private val BASE32: BaseNCodec = Base32()
		private val BASE32_ENCRYPTED_NAME_PATTERN = Pattern.compile("^(0|1S)?(([A-Z2-7]{8})*[A-Z2-7=]{8})$")
	}
}
